android/src/main/java/com/polidea/flutter_ble_lib/delegate/DiscoveryDelegate.java

```diff
@@ -104,87 +106,57 @@ public void onError(BleError error) {
-
-                adapter.getCharacteristicsForService(foundService.getId(),
-                        new OnSuccessCallback<Characteristic[]>() {
-                            @Override
-                            public void onSuccess(Characteristic[] data) {
-                                CharacteristicsResponse characteristicsResponse = new CharacteristicsResponse(data, finalService);
-                                try {
-                                    String json = new CharacteristicsResponseJsonConverter().toJson(characteristicsResponse);
-                                    result.success(json);
-                                } catch (JSONException e) {
-                                    result.error(e.toString(), e.getMessage(), null);
-                                }
-                            }
-                        },
-                        errorCallback);
+            if (foundService == null) {
+                result.error("UnknownServiceException", "Service not found", "Unknown service UUID " + serviceUuid);
+                return;
             }
-        };
 
-        adapter.getServicesForDevice(deviceId, serviceFetchSuccessCallback, errorCallback);
+            Characteristic[] characteristics = adapter.getCharacteristicsForService(foundService.getId());
+            MultiCharacteristicsResponse characteristicsResponse = new MultiCharacteristicsResponse(characteristics, foundService);
+            String json = multiCharacteristicsResponseJsonConverter.toJson(characteristicsResponse);
+            result.success(json);
+        } catch (BleError error) {
+            error.printStackTrace();
+            result.error(String.valueOf(error.errorCode.code), error.reason, bleErrorJsonConverter.toJson(error));
+        } catch (JSONException e) {
+            e.printStackTrace();
+            result.error(null, e.getMessage(), null);
+        }
     }
 
     private void getServices(String deviceId, final MethodChannel.Result result) {
-        adapter.getServicesForDevice(deviceId,
-                new OnSuccessCallback<Service[]>() {
-                    @Override
-                    public void onSuccess(final Service[] data) {
-                        try {
-                            result.success(new ServiceJsonConverter().toJson(data));
-                        } catch (JSONException e) {
-                            result.error(e.toString(), e.getMessage(), null);
-                        }
-                    }
-                }, new OnErrorCallback() {
-                    @Override
-                    public void onError(BleError error) {
-                        result.error(String.valueOf(error.errorCode.code), error.reason, bleErrorJsonConverter.toJson(error));
-                    }
-                });
+        try {
+            Service[] services = adapter.getServicesForDevice(deviceId);
+            result.success(serviceJsonConverter.toJson(services));
```

Comment:
Can services be null/empty and not trigger an exception? If yes, what should happen?

Reply:
Services can be empty and cannot be null. After the discussion with @Cierpliwy we decided that this is an expected behavior.